In [1]:
import pandas as pd

# **Loading the file in dataframe**

In [2]:
df=pd.read_csv('/content/Day10_Flight_Operations_Dataset.csv')

# **Understanding the dataset**

In [11]:
print("Shape : ",df.shape)
print("\n\nBasic Infomation :")
df.info()
print("\n\nColumns:")
print(df.columns.tolist())
print("\n\nStatistical summary")
df.describe(include='all').T

Shape :  (180, 17)


Basic Infomation :
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Av

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Flight_ID,180,180,FL0001,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Flight_Date,180,111,2026-02-10,6,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Airline,180,6,Vistara,37,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Origin,180,11,Delhi,43,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Destination,180,8,Delhi,51,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Aircraft,180,4,Airbus A319,60,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Travel_Class,180,3,Economy,134,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Passengers,180.0,NaN,NaN,NaN,122.772222,45.635651,45.0,78.75,125.0,162.25,210.0
Seat_Capacity,180.0,NaN,NaN,NaN,197.238889,15.408176,180.0,186.0,189.0,210.0,220.0
Average_Ticket_Price,180.0,NaN,NaN,NaN,6111.677778,3615.22004,1949.0,3815.5,5380.0,6790.0,27832.0


# **Checking Data Quality**

In [13]:
print("Missing values:",df.isnull().sum())
print("Duplicate rows:", df.duplicated().sum())

Missing values: Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             7
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64
Duplicate rows: 0


# **Data Transformation**

In [14]:
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'])
df['Month'] = df['Flight_Date'].dt.month
df['Day_Name'] = df['Flight_Date'].dt.day_name()
df['Estimated_Revenue'] = df['Passengers'] * df['Average_Ticket_Price']
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Month,Day_Name,Estimated_Revenue
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,3,Monday,1447740
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,6,Sunday,312760
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,5,Tuesday,315771
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,5,Saturday,580023
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,2,Friday,1157890


# **Selection and Filtering**

In [15]:
# Select important columns
important_columns = df[
    ['Flight_ID', 'Airline', 'Origin', 'Destination',
     'Passengers', 'Delay_Minutes', 'Flight_Status']
]
important_columns.head()

,Flight_ID,Airline,Origin,Destination,Passengers,Delay_Minutes,Flight_Status
0,FL0001,SpiceJet,Mumbai,Bengaluru,210,50.0,Delayed
1,FL0002,Akasa Air,Delhi,Mumbai,70,0.0,On Time
2,FL0003,Air India,Kochi,Delhi,67,5.0,On Time
3,FL0004,Akasa Air,Srinagar,Delhi,153,35.0,Delayed
4,FL0005,Air India,Pune,Delhi,178,20.0,Delayed


In [16]:
#Flights delayed by more than 30 minutes
high_delay = df[df['Delay_Minutes'] > 30]
print("Flights delayed by more than 30 minutes:", len(high_delay))
high_delay[['Flight_ID', 'Airline', 'Delay_Minutes', 'Weather']].head()

Flights delayed by more than 30 minutes: 40


,Flight_ID,Airline,Delay_Minutes,Weather
0,FL0001,SpiceJet,50.0,Storm
3,FL0004,Akasa Air,35.0,Storm
5,FL0006,Akasa Air,35.0,Cloudy
8,FL0009,Akasa Air,75.0,Cloudy
21,FL0022,Vistara,75.0,Clear


# **Sorting Records**

In [17]:
# Sort by highest delay
df.sort_values(by='Delay_Minutes', ascending=False)[
    ['Flight_ID', 'Airline', 'Delay_Minutes', 'Weather']
].head(10)

,Flight_ID,Airline,Delay_Minutes,Weather
26,FL0027,SpiceJet,110.0,Clear
177,FL0178,IndiGo,110.0,Fog
82,FL0083,Air India Express,110.0,Cloudy
57,FL0058,IndiGo,110.0,Cloudy
50,FL0051,Air India,110.0,Rain
128,FL0129,SpiceJet,110.0,Cloudy
146,FL0147,IndiGo,110.0,Rain
169,FL0170,IndiGo,110.0,Rain
161,FL0162,Air India,110.0,Rain
21,FL0022,Vistara,75.0,Clear


# **Grouping and Aggregation**

In [18]:
# Performance by airline
airline_analysis = df.groupby('Airline').agg(
    Total_Flights=('Flight_ID', 'count'),
    Total_Passengers=('Passengers', 'sum'),
    Average_Delay=('Delay_Minutes', 'mean'),
    Total_Revenue=('Estimated_Revenue', 'sum'),
    Average_Satisfaction=('Passenger_Satisfaction', 'mean')
).sort_values(by='Total_Revenue', ascending=False)

airline_analysis

,Total_Flights,Total_Passengers,Average_Delay,Total_Revenue,Average_Satisfaction
Airline,,,,,
Vistara,37,4484,18.628571,28689467,3.378378
IndiGo,32,4158,25.031250,25709131,3.343750
SpiceJet,32,3929,24.137931,24195935,3.531250
Air India Express,24,3027,11.291667,23944164,3.666667
Akasa Air,31,3427,14.400000,18872059,3.419355
Air India,24,3074,21.260870,16232418,3.750000


# **Meaningful Questions and Answers**

In [19]:
# Q1. Which airline generated the highest estimated revenue?
print(airline_analysis['Total_Revenue'].idxmax())
print(airline_analysis['Total_Revenue'].max())

Vistara
28689467


In [20]:
# Q2. Which airline has the lowest average delay?
print(airline_analysis['Average_Delay'].idxmin())
print(airline_analysis['Average_Delay'].min())

Air India Express
11.291666666666666


In [22]:
# Q7. What percentage of flights were on time?
on_time= ((df['Flight_Status'] == 'On Time').mean() * 100)
print(f"On-time flights: {on_time:.2f}%")

On-time flights: 63.33%


# **Key Observations**

Based on the analysis:

1. **Vistara generated the highest total estimated revenue**, indicating strong overall financial performance in this dataset.
2. **Air India generated the lowest total estimated revenue**
3. **Air India Express had the lowest average delay**
4. **IndiGo had a relatively higher average delay** among the airlines.
5. **On-time flights were the dominant flight status**, accounting for approximately **63% of all flights** in the dataset.
